### PR TITLE
Review and edit security documentation

### PR DESCRIPTION
**Pull request by willingc**
This PR edits the two existing documents on security for JupyterHub.

- Security settings document now contains only information about the items that are typically configured for security (SSL, cookie secret, proxy authorization token)
- Security settings document was changed to Restructured Text for easier linking and nicer information boxes/alerts
- The Security overview document focuses more on best practices and how JupyterHub was designed for semi-trusted users

[Rendered Security settings doc](http://test-jupyterhub.readthedocs.io/en/deduplicate/security-basics.html)

[Security overview doc](http://test-jupyterhub.readthedocs.io/en/deduplicate/websecurity.html)

### CODE REVIEW

File: docs/source/security-basics.rst [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+configuration settings are the main aspects of security configuration:
+
+1. :ref:`SSL encryption <ssl-encryption>` (to enable HTTPS)
+2. :ref:`Cookie secret <cookie-secret>` (a key for encrypting browser cookies)
+3. Proxy :ref:`authentication token <authentication-token>` (used for the Hub and
+   other services to authenticate to the Proxy)
+
+The Hub hashes all secrets (e.g., auth tokens) before storing them in its
+database. A loss of control over read-access to the database should have no
+security impact on your deployment.
```
Comment (minrk) [inline docs/source/security-basics.rst]: "no impact" might be an overstatement. Maybe rephrase to "minimal impact, but if your database has been compromised, it is still a good idea to revoke existing tokens."

File: docs/source/security-basics.rst [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+Generating and storing as a cookie secret file
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The cookie secret should be 32 random bytes, encoded as hex, and is typically
+stored in a ``cookie_secret`` file. An example command to generate the
+``cookie_secret`` file is:
+
+.. code-block:: bash
+
+    openssl rand -hex 32 > /srv/jupyterhub/cookie_secret
```
Comment (minrk) [inline docs/source/security-basics.rst]: The default name is `jupyterhub_cookie_secret`, so probably use that.

File: docs/source/security-basics.rst [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+Default if token is not set
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you don't set the Proxy authentication token, the Hub will generate a random
+key itself, which means that any time you restart the Hub you **must also
+restart the Proxy**. If the proxy is a subprocess of the Hub, this should happen
+automatically (this is the default configuration).
+
+Setting a token when using services
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (minrk) [inline docs/source/security-basics.rst]: I think we can remove this last section. nbgrader operating as a service no longer needs to talk directly to the proxy, and in general exposing the proxy as somethign for things other than JupyterHub to talk to is something we shouldn't do anymore now that there are multiple proxy implementations.

File: docs/source/websecurity.md [hunk 4]
```diff
@@ -1,80 +1,116 @@
...
 
-### Disabling user config
+### Unavailable subdomains
 
-If subdomains are not available or not desirable, 0.5 also adds an option
-`Spawner.disable_user_config`, which you can set to prevent the user-owned
-configuration files from being loaded. This leaves only package installation
-and PATHs as things the admin must enforce.
+When subdomains are not available or not desirable, three steps can be taken
+to secure JupyterHub from untrusted users.
```
Comment (minrk) [inline docs/source/websecurity.md]: I don't think we should be claiming that disable_user_config and other options are for use in case subdomains are unavailable. These are all still valid if subdomains are in use, their consequences are just less severe.

File: docs/source/websecurity.md [hunk 5]
```diff
@@ -1,80 +1,116 @@
...
+#### Prevent spawners from evaluating shell configuration files
+
+For most Spawners, `PATH` is not something users can influence, but care should
+be taken to ensure that the Spawner does *not* evaluate shell configuration
 files prior to launching the server.
 
+#### Isolate packages using virtualenv
+
 Package isolation is most easily handled by running the single-user server in
 a virtualenv with disabled system-site-packages.
```
Comment (minrk) [inline docs/source/websecurity.md]: Add "The user should not have permission to install packages into this environment."

**Comment by willingc** [comment]
Would you mind rereading the cookie secret edits? Thanks @minrk.

**Comment by minrk** [comment]
Looks great, thanks!
